### PR TITLE
Release/babylon 1.4.0 - > main

### PR DESCRIPTION
**Pull request by PawelPawelec-RDX**

### CODE REVIEW

**Review by krzlabrdx — APPROVED** [review]

**Review by krzlabrdx — APPROVED** [review]

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 1]
```diff
@@ -879,11 +879,99 @@ tags:
             }
       ```
 
+      ### Iterating over all key value store keys
+
+      You can use `/state/key-value-store/keys` endpoint to iterate over all keys in certain key value store.
```
Comment (dhedey) [inline src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml]: Minor English article corrections:
```
      You can use the `/state/key-value-store/keys` endpoint to iterate over all the keys in a certain key value store.
```

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 2]
```diff
@@ -879,11 +879,99 @@ tags:
...
+      If the queried key value store contains more than one page of keys, a next_cursor will be returned, which can be used as cursor in the request to fetch further pages.
```
Comment (dhedey) [inline src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml]: Minor English article corrections:
```
If the queried key value store contains more than one page of keys, a `next_cursor` will be returned, which can be used as a cursor in the request to fetch further pages.
```

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 3]
```diff
@@ -879,11 +879,99 @@ tags:
...
+      i.e:
+      ```
+      /state/key-value-store/keys
+
+      {
+        "key_value_store_address": "internal_keyvaluestore_tdx_2_1kzjd929eqlzd9n02uuj8jd48705vcrpvhv4mgxnaltrgystnca3qxk"
+      }
+      ```
+
+      Will respond with list of all keys (in both `hex` and `programmatic_json` form) and state version when they were last updated.
```
Comment (dhedey) [inline src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:896]: Minor English article corrections:

```
      Will respond with a list of all the keys (in both `hex` and `programmatic_json` form) and the state version when they were last updated.
```

However(!) I'd suggest we say it's actually not all keys - it's importantly the *first page of keys*, and we should explain how it's ordered too.

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 4]
```diff
@@ -879,11 +879,99 @@ tags:
...
+
```
Comment (dhedey) [inline src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:897]: Also - just to verify - if a key value store entry is **deleted**, what happens?

I think it's still returned by this API? Or do we filter them out?

If it doesn't filter them out, we should explain that this API doesn't do so, and explain how the entry read endpoint handles us looking up a key of an entry that's been deleted.

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 5]
```diff
@@ -879,11 +879,99 @@ tags:
...
+          ...
+        ],
+        "key_value_store_address": "internal_keyvaluestore_tdx_2_1kzjd929eqlzd9n02uuj8jd48705vcrpvhv4mgxnaltrgystnca3qxk"
+      }
+      ```
+
       ### Querying for the content under specific keys
 
-      After that you can use `/state/key-value-store/data` endpoint to query the content for given keys. To do that you can either use key in form of `json` or `hex`.
+      You can either rely on keys returned from `/state/key-value-store/keys` endpoint or construct keys in json form on your own. It's possible if you created that key value store and you know what is the form of keys.
```
Comment (dhedey) [inline src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml]: Suggested:
```
You can either use the keys returned from the `/state/key-value-store/keys` endpoint or construct keys in json form on your own, if you know the key type of the `KeyValueStore`.
```

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 6]
```diff
@@ -879,11 +879,99 @@ tags:
...
+
+      #### Querying using hex keys
+
+      Simply grab `raw_hex` key returned from `/state/key-value-store/keys` endpoint and call `/state/key-value-store/data` with it as parameter.
```
Comment (dhedey) [inline src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:925]:
```
      Simply grab the `raw_hex` key returned from the `/state/key-value-store/keys` endpoint and call `/state/key-value-store/data` with it as a parameter.
```

**Review by dhedey — COMMENTED** [review]
Awesome, thanks Pawel!

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml (hunk 3 above)
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:896]: how about

```
Will respond with a first page of all keys (in both `hex` and `programmatic_json` form) and the state version when they were last updated. Returned keys are ordered by their last change, first key on a list will be one that was most recently updated/created. 
```

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml (hunk 4 above)
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:897]: we filter out deleted keys. do we want to document that somehow?

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml (hunk 3 above)
Comment (dhedey) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:896]: I suggest this tweak, if you're happy with it? 
```
Will respond with a first page of the keys of the current entries of the key value store. This collection is ordered by the state version of the entry's last update, descending. In other words, the first entry on the list will be one that was most recently updated/created. Each item includes the entry's key (in both `hex` and `programmatic_json` form) and the state version when it was last updated.
```

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml (hunk 4 above)
Comment (dhedey) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:897]: I think if we rewrite as per my other message, and say "current entries", this implies deleted keys aren't returned.

**Pull request merged**